syntax = "proto3";

service RAFT {
	rpc SayHello (HelloRequest) returns (HelloReply) {}
	rpc RequestVote (RequestVoteRequest) returns (RequestVoteReply) {}
	rpc AppendEntries (AppendEntriesRequest) returns (AppendEntriesReply) {}
	rpc Put (PutRequest) returns (PutReply) {}
	rpc Get (GetRequest) returns (GetReply) {}
}

message HelloRequest {
	string name = 1;
}

message HelloReply {
	string message = 1;
}

message Entry {
	int32 term = 1;
	string key = 2;
	string value = 3;
}

message AppendEntriesRequest {
	int32 term = 1;
	string leaderId = 2;
	int32 prevLogIndex = 3;
	int32 prevLogTerm = 4;
	repeated Entry entries = 5;
	int32 leaderCommit = 6;
}

message AppendEntriesReply {
	int32 term = 1;
	bool success = 2;
	string receiver = 3;
	int32 cnt = 4;
}

message RequestVoteRequest {
	int32 term = 1;
	string candidateId = 2;
	int32 lastLogIndex = 3;
	int32 lastLogTerm = 4;
}

message RequestVoteReply {
	int32 term = 1;
	bool voteGranted = 2;
}

message PutRequest {
	string key = 1;
	string value = 2;
}

message PutReply {
	bool success = 1;
}

message GetRequest {
	string key = 1;
}

message GetReply {
	bool success = 1;
	string value = 2;
}